Discrete Search Improvements

piero2c opened the pull request:
# Pull Request Template

## Features

* Replaces `Objective` and `AsyncObjective` abstractions with the new `SyncEvaluator` and `AsyncEvaluator`
* New `SearchObjectives` abstraction to handle evaluation caching, cheap/expensive objectives and search constraints.
* New search algorithms:
  * Random Search
  * Local Search
  * Regularized Evolution
* Better logging and signature for search algorithms
* Tests for all new features
* Basic guide tutorials update 

## Changes

Please delete options that are not relevant.

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (causes existing functionality to not work as expected)
- [X] Documentation update

## Final checks:

- [X] Follows guidelines of project
- [X] Self-review of code
- [X] Commented code in hard-to-understand areas
- [X] Corresponding changes to the documentation
- [X] Changes generate no new warnings
- [X] Dependent changes have already been merged
- [X] Checked code correction any misspellings

gugarosa commented on archai/nlp/objectives/parameters.py:
```diff
@@ -39,24 +30,21 @@ def evaluate(self, model: ArchaiModel, dataset: DatasetProvider, budget: Optiona
 
-        """
+        Args:
+            exclude_cls (Optional[List[nn.Module]], optional): List of PyTorch module classes
```
Do you think that we should use annotated typing on the docstrings?

I have seen on this guide that it may not be necessary since we are already annotating the headers: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

```
def function_with_pep484_type_annotations(param1: int, param2: str) -> bool:
    """Example function with PEP 484 type annotations.

    Args:
        param1: The first parameter.
        param2: The second parameter.

    Returns:
        The return value. True for success, False otherwise.

    """
```

piero2c replied:
That's a good question. I like having them there too, but they can be tricky to deal with when the type expressions are lengthy...

gugarosa replied:
That's true!

I will leave it up to you, please feel free to choose the style that we will be using. Just please let me know, so I can make the proper changes on other files 😄

gugarosa commented:
Regarding the conflicts, all of them are because `archai.datasets` have been moved into `archai.cv.datasets`, just to keep the same "structure" that we have for the `nlp` package.

PS: We can always change that in the future though!

piero2c commented:
Updated PR:
* New naming convention for evaluators and SearchObjective methods
* YAML export as default for ConfigSearchSpace
* Updated docstrings
* Updated tutorials
* Updated torch profiler code
